### PR TITLE
Use `add_link_options` and `target_link_options` in cmake

### LINKED ISSUE #92843: Remove the `add_linker_flag` function and instead use `target_link_options`

### PR DESCRIPTION
**Pull request by jtschuster**
Fixes https://github.com/dotnet/runtime/issues/92843

Replaces `add_linker_flag` with `add_linker_options` and `target_linker_options`, where possible.

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
There's a slight UX difference between the `add_linker_flag` function we have and `add_link_options`. Our function takes the config as a second parameter. `add_link_options` doesn't, so we'd need to use generator expressions.

Also the way you specify the `-Wl,` flags is different as well. Here's a table that shows how to go from the `add_linker_flag` way to the `add_link_options` way:

| add_linker_flag | add_link_options | Useful doc link|
| ----- | -----|-----|
| `add_linker_flag(foo DEBUG)` | `add_link_options($<$<CONFIG:Debug>:foo>)` | https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html |
| `add_linker_flag(-Wl,-z,notext)` | `add_link_options("LINKER:-z,notext")` | https://cmake.org/cmake/help/latest/command/add_link_options.html#handling-compiler-driver-differences |

File: eng/native/configurecompiler.cmake [hunk 1]
```diff
@@ -117,27 +117,27 @@ if (MSVC)
     # Debug build specific flags
     # The Ninja generator doesn't appear to have the default `/INCREMENTAL:ON` that
     # the Visual Studio generator has. Therefore we will override the default for Visual Studio only.
-    add_link_options(/INCREMENTAL:NO DEBUG)
-    add_link_options(/OPT:REF DEBUG)
-    add_link_options(/OPT:NOICF DEBUG)
+    add_link_options($<$<CONFIG:DEBUG>:LINKER:/INCREMENTAL:NO>)
```
Comment (jkoritzinsky) [inline eng/native/configurecompiler.cmake:120]: We should only need the `LINKER:` prefix for arguments that were passed with `-Wl,` beforehand. Were you seeing errors without it for the other arguments?

**Comment by jkoritzinsky** [comment]
The remaining failures on Windows look like they're in the RC-only targets. Right now these targets remove linker flags that they are not compatible with. Instead, we can update the `add_link_options` for these flags to exclude RC-only targets by using the generator expression: `$<$<NOT:$<LINK_LANGUAGE:RC>>:flag>`. Looks like we need this for the `/LTCG` flag and the `/CETCOMPAT` flag.

File: eng/native/configurecompiler.cmake [hunk 2]
```diff
@@ -286,28 +286,28 @@ endif()
...
 if(CLR_CMAKE_HOST_LINUX)
   add_compile_options($<$<COMPILE_LANGUAGE:ASM>:-Wa,--noexecstack>)
-  add_linker_flag(-Wl,--build-id=sha1)
-  add_linker_flag(-Wl,-z,relro,-z,now)
+  add_link_options("LINKER:--build-id=sha1")
+  add_link_options("LINKER:-z,relro,-z,now")
 elseif(CLR_CMAKE_HOST_FREEBSD)
   add_compile_options($<$<COMPILE_LANGUAGE:ASM>:-Wa,--noexecstack>)
-  add_linker_flag("-Wl,--build-id=sha1")
+  add_link_options("LINKER--build-id=sha1")
```
Comment (jkoritzinsky) [inline eng/native/configurecompiler.cmake:299]:
```suggestion
  add_link_options("LINKER:--build-id=sha1")
```

File: eng/native/configurecompiler.cmake [hunk 3]
```diff
@@ -166,8 +155,7 @@ if (CLR_CMAKE_ENABLE_SANITIZERS)
       # /RTC1 is added by default by CMake and incompatible with ASAN, so remove it.
       string(REPLACE "/RTC1" "" CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG}")
       string(REPLACE "/RTC1" "" CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG}")
-      string(REPLACE "/RTC1" "" CMAKE_SHARED_LINKER_FLAGS_DEBUG "${CMAKE_SHARED_LINKER_FLAGS_DEBUG}")
-      string(REPLACE "/RTC1" "" CMAKE_EXE_LINKER_FLAGS_DEBUG "${CMAKE_EXE_LINKER_FLAGS_DEBUG}")
+      string(REPLACE "/RTC1" $<IF:$<CONFIG:DEBUG>,"","/RTC1"> LINK_OPTIONS "${LINK_OPTIONS}")
```
Comment (jkoritzinsky) [inline eng/native/configurecompiler.cmake:158]: I never thought about using a generator expression in a replacement here. Interesting! However, we don't need to keep `/RTC1` on all platforms (it's only added-by-default in DEBUG). Also, this option is set specifically in the old-style flags. Let's leave the changes to this flag as-is.

File: eng/native/ijw/IJW.cmake [hunk 4]
```diff
@@ -65,8 +65,7 @@ if (CLR_CMAKE_HOST_WIN32)
   set_directory_properties(PROPERTIES COMPILE_OPTIONS "${dirCompileOptions}")
 
   # IJW tests needs to load DLLs from somewhere other than System32
-  string(REPLACE "/DEPENDENTLOADFLAG:0x800" "" CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS}")
-  string(REPLACE "/DEPENDENTLOADFLAG:0x800" "" CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS}")
+  string(REPLACE "/DEPENDENTLOADFLAG:0x800" "" LINK_OPTIONS "${LINK_OPTIONS}")
```
Comment (jkoritzinsky) [inline eng/native/ijw/IJW.cmake:68]: You probably need to use `get_directory_property` and `set_directory_properties` here to get this to work right.

File: eng/native/configurecompiler.cmake [hunk 5]
```diff
@@ -112,13 +112,14 @@ if (MSVC)
 
   if (CMAKE_GENERATOR MATCHES "^Visual Studio.*$")
     # Debug build specific flags
-    # The Ninja generator doesn't appear to have the default `/INCREMENTAL:ON` that
+    # The Ninja generator doesn't appear to have the default `/INCREMENTAL:NO` that
```
Comment (jkoritzinsky) [inline eng/native/configurecompiler.cmake:115]: The previous wording is correct.

```suggestion
    # The Ninja generator doesn't appear to have the default `/INCREMENTAL:ON` that
```

Comment (jtschuster) [reply eng/native/configurecompiler.cmake:115]: I don't see any `ON` option in the documentation https://learn.microsoft.com/en-us/cpp/build/reference/incremental-link-incrementally?view=msvc-170, should it just be `/INCREMENTAL` then?

Comment (jkoritzinsky) [reply eng/native/configurecompiler.cmake:115]: You're right

```suggestion
    # The Ninja generator doesn't appear to have the default `/INCREMENTAL` that
```

**Review by jkoritzinsky — APPROVED** [review]

**Comment by jtschuster** [comment]
Failure looks like https://github.com/dotnet/runtime/issues/93527

**Review by jkoritzinsky — APPROVED** [review]

**Comment by lewing** [comment]
Looks like this broke the official build 

```
2023-10-18T11:44:14.3256632Z   CMake Error at pgosupport.cmake:21 (target_link_options):
2023-10-18T11:44:14.3262066Z     target_link_options called with invalid arguments
2023-10-18T11:44:14.3262652Z   Call Stack (most recent call first):
2023-10-18T11:44:14.3263392Z     jit/CMakeLists.txt:626 (add_pgo)
2023-10-18T11:44:14.3264832Z   
2023-10-18T11:44:14.3266461Z   
2023-10-18T11:44:14.3268303Z   CMake Error at pgosupport.cmake:22 (target_link_options):
2023-10-18T11:44:14.3269877Z     target_link_options called with invalid arguments
2023-10-18T11:44:14.3271707Z   Call Stack (most recent call first):
2023-10-18T11:44:14.3273617Z     jit/CMakeLists.txt:626 (add_pgo)
2023-10-18T11:44:14.3274400Z   
2023-10-18T11:44:14.3275848Z   
2023-10-18T11:44:15.0339628Z   Read file version from native version header at 'D:/a/_work/1/s/artifacts/obj/_version.h'.
2023-10-18T11:44:15.0436800Z   CMake Error at pgosupport.cmake:21 (target_link_options):
2023-10-18T11:44:15.0437813Z     target_link_options called with invalid arguments
2023-10-18T11:44:15.0438984Z   Call Stack (most recent call first):
2023-10-18T11:44:15.0441399Z     dlls/mscoree/coreclr/CMakeLists.txt:215 (add_pgo)
2023-10-18T11:44:15.0443545Z   
2023-10-18T11:44:15.0445423Z   
2023-10-18T11:44:15.0447812Z   CMake Error at pgosupport.cmake:22 (target_link_options):
2023-10-18T11:44:15.0449982Z     target_link_options called with invalid arguments
2023-10-18T11:44:15.0451924Z   Call Stack (most recent call first):
2023-10-18T11:44:15.0454038Z     dlls/mscoree/coreclr/CMakeLists.txt:215 (add_pgo)
2023-10-18T11:44:15.0456001Z   
2023-10-18T11:44:15.0457863Z   
2023-10-18T11:44:15.0841406Z   -- Configuring incomplete, errors occurred!
2023-10-18T11:44:15.1132740Z ##[error]BUILD: Error: failed to generate native component build project
```

**Comment by janvorli** [comment]
@jtschuster, @jkoritzinsky, @jkotas it also accidentally enabled CETCOMPAT for executables (e.g. corerun.exe).

**Pull request merged**